In [1]:
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("TradeCorpETL").getOrCreate()
spark.version

'3.5.0'

In [2]:
df_categories = spark.read.csv("../data/categories.csv",header=True,inferSchema=True)

In [3]:
df_categories.show()

+-----------+--------------+--------------------+-------+
|category_id| category_name|         description|picture|
+-----------+--------------+--------------------+-------+
|          1|     Beverages|Soft drinks, coff...|     \x|
|          2|    Condiments|Sweet and savory ...|     \x|
|          3|   Confections|Desserts, candies...|     \x|
|          4|Dairy Products|             Cheeses|     \x|
|          5|Grains/Cereals|Breads, crackers,...|     \x|
|          6|  Meat/Poultry|      Prepared meats|     \x|
|          7|       Produce|Dried fruit and b...|     \x|
|          8|       Seafood|    Seaweed and fish|     \x|
+-----------+--------------+--------------------+-------+



In [4]:
df_customers = spark.read.csv("../data/customers.csv",header=True,inferSchema=True)

In [5]:
df_customers.show()

+-----------+--------------------+------------------+--------------------+--------------------+------------+------+-----------+-----------+--------------+--------------+
|customer_id|        company_name|      contact_name|       contact_title|             address|        city|region|postal_code|    country|         phone|           fax|
+-----------+--------------------+------------------+--------------------+--------------------+------------+------+-----------+-----------+--------------+--------------+
|      ALFKI| Alfreds Futterkiste|      Maria Anders|Sales Representative|       Obere Str. 57|      Berlin|  NULL|      12209|    Germany|   030-0074321|   030-0076545|
|      ANATR|Ana Trujillo Empa...|      Ana Trujillo|               Owner|Avda. de la Const...| M�xico D.F.|  NULL|      05021|     Mexico|  (5) 555-4729|  (5) 555-3745|
|      ANTON|Antonio Moreno Ta...|    Antonio Moreno|               Owner|     Mataderos  2312| M�xico D.F.|  NULL|      05023|     Mexico|  (5) 555-3

In [6]:
df_employees = spark.read.csv("../data/employees.csv",header=True,inferSchema=True)

In [7]:
df_employees.show()

+-----------+---------+----------+--------------------+-----------------+----------+----------+--------------------+--------+------+-----------+-------+--------------+---------+-----+--------------------+----------+--------------------+
|employee_id|last_name|first_name|               title|title_of_courtesy|birth_date| hire_date|             address|    city|region|postal_code|country|    home_phone|extension|photo|               notes|reports_to|          photo_path|
+-----------+---------+----------+--------------------+-----------------+----------+----------+--------------------+--------+------+-----------+-------+--------------+---------+-----+--------------------+----------+--------------------+
|          1|  Davolio|     Nancy|Sales Representative|              Ms.|1948-12-08|1992-05-01|507 - 20th Ave. E...| Seattle|    WA|      98122|    USA|(206) 555-9857|     5467|   \x|Education include...|         2|http://accweb/emm...|
|          2|   Fuller|    Andrew|Vice President, S.

In [8]:
df_order_details = spark.read.csv("../data/order_details.csv",header=True,inferSchema=True)

In [9]:
df_order_details.show()

+--------+----------+----------+--------+--------+
|order_id|product_id|unit_price|quantity|discount|
+--------+----------+----------+--------+--------+
|   10248|        11|      14.0|      12|     0.0|
|   10248|        42|       9.8|      10|     0.0|
|   10248|        72|      34.8|       5|     0.0|
|   10249|        14|      18.6|       9|     0.0|
|   10249|        51|      42.4|      40|     0.0|
|   10250|        41|       7.7|      10|     0.0|
|   10250|        51|      42.4|      35|    0.15|
|   10250|        65|      16.8|      15|    0.15|
|   10251|        22|      16.8|       6|    0.05|
|   10251|        57|      15.6|      15|    0.05|
|   10251|        65|      16.8|      20|     0.0|
|   10252|        20|      64.8|      40|    0.05|
|   10252|        33|       2.0|      25|    0.05|
|   10252|        60|      27.2|      40|     0.0|
|   10253|        31|      10.0|      20|     0.0|
|   10253|        39|      14.4|      42|     0.0|
|   10253|        49|      16.0

In [10]:
df_orders = spark.read.csv("../data/orders.csv",header=True,inferSchema=True)

In [11]:
df_orders.show()

+--------+-----------+-----------+----------+-------------+------------+--------+-------+--------------------+--------------------+--------------+-----------+----------------+------------+
|order_id|customer_id|employee_id|order_date|required_date|shipped_date|ship_via|freight|           ship_name|        ship_address|     ship_city|ship_region|ship_postal_code|ship_country|
+--------+-----------+-----------+----------+-------------+------------+--------+-------+--------------------+--------------------+--------------+-----------+----------------+------------+
|   10248|      VINET|          5|1996-07-04|   1996-08-01|  1996-07-16|       3|  32.38|Vins et alcools C...|  59 rue de l'Abbaye|         Reims|       NULL|           51100|      France|
|   10249|      TOMSP|          6|1996-07-05|   1996-08-16|  1996-07-10|       1|  11.61|  Toms Spezialit�ten|       Luisenstr. 48|       M�nster|       NULL|           44087|     Germany|
|   10250|      HANAR|          4|1996-07-08|   1996-08

In [12]:
df_products = spark.read.csv("../data/products.csv",header=True,inferSchema=True)

In [13]:
df_products.show()

+----------+--------------------+-----------+-----------+--------------------+----------+--------------+--------------+-------------+------------+
|product_id|        product_name|supplier_id|category_id|   quantity_per_unit|unit_price|units_in_stock|units_on_order|reorder_level|discontinued|
+----------+--------------------+-----------+-----------+--------------------+----------+--------------+--------------+-------------+------------+
|         1|                Chai|          8|          1|  10 boxes x 30 bags|      18.0|            39|             0|           10|           1|
|         2|               Chang|          1|          1|  24 - 12 oz bottles|      19.0|            17|            40|           25|           1|
|         3|       Aniseed Syrup|          1|          2| 12 - 550 ml bottles|      10.0|            13|            70|           25|           0|
|         4|Chef Anton's Caju...|          2|          2|      48 - 6 oz jars|      22.0|            53|             0

In [14]:
df_shippers = spark.read.csv("../data/shippers.csv",header=True,inferSchema=True)

In [15]:
df_shippers.show()

+----------+-----------------+--------------+
|shipper_id|     company_name|         phone|
+----------+-----------------+--------------+
|         1|   Speedy Express|(503) 555-9831|
|         2|   United Package|(503) 555-3199|
|         3| Federal Shipping|(503) 555-9931|
|         4|Alliance Shippers|1-800-222-0451|
|         5|              UPS|1-800-782-7892|
|         6|              DHL|1-800-225-5345|
+----------+-----------------+--------------+



In [16]:
df_suppliers = spark.read.csv("../data/suppliers.csv",header=True,inferSchema=True)

In [17]:
df_suppliers.show()

+-----------+--------------------+--------------------+--------------------+--------------------+-----------+--------+-----------+---------+---------------+---------------+--------------------+
|supplier_id|        company_name|        contact_name|       contact_title|             address|       city|  region|postal_code|  country|          phone|            fax|            homepage|
+-----------+--------------------+--------------------+--------------------+--------------------+-----------+--------+-----------+---------+---------------+---------------+--------------------+
|          1|      Exotic Liquids|    Charlotte Cooper|  Purchasing Manager|      49 Gilbert St.|     London|    NULL|    EC1 4SD|       UK| (171) 555-2222|           NULL|                NULL|
|          2|New Orleans Cajun...|       Shelley Burke| Order Administrator|      P.O. Box 78934|New Orleans|      LA|      70117|      USA| (100) 555-4822|           NULL|         #CAJUN.HTM#|
|          3|Grandma Kelly's H

In [18]:
df_suppliers.printSchema()

root
 |-- supplier_id: integer (nullable = true)
 |-- company_name: string (nullable = true)
 |-- contact_name: string (nullable = true)
 |-- contact_title: string (nullable = true)
 |-- address: string (nullable = true)
 |-- city: string (nullable = true)
 |-- region: string (nullable = true)
 |-- postal_code: string (nullable = true)
 |-- country: string (nullable = true)
 |-- phone: string (nullable = true)
 |-- fax: string (nullable = true)
 |-- homepage: string (nullable = true)



In [19]:
df_shippers.printSchema()

root
 |-- shipper_id: integer (nullable = true)
 |-- company_name: string (nullable = true)
 |-- phone: string (nullable = true)



In [20]:
df_products.printSchema()

root
 |-- product_id: integer (nullable = true)
 |-- product_name: string (nullable = true)
 |-- supplier_id: integer (nullable = true)
 |-- category_id: integer (nullable = true)
 |-- quantity_per_unit: string (nullable = true)
 |-- unit_price: double (nullable = true)
 |-- units_in_stock: integer (nullable = true)
 |-- units_on_order: integer (nullable = true)
 |-- reorder_level: integer (nullable = true)
 |-- discontinued: integer (nullable = true)



In [21]:
df_orders.printSchema()

root
 |-- order_id: integer (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- employee_id: integer (nullable = true)
 |-- order_date: date (nullable = true)
 |-- required_date: date (nullable = true)
 |-- shipped_date: date (nullable = true)
 |-- ship_via: integer (nullable = true)
 |-- freight: double (nullable = true)
 |-- ship_name: string (nullable = true)
 |-- ship_address: string (nullable = true)
 |-- ship_city: string (nullable = true)
 |-- ship_region: string (nullable = true)
 |-- ship_postal_code: string (nullable = true)
 |-- ship_country: string (nullable = true)



In [22]:
df_order_details.printSchema()

root
 |-- order_id: integer (nullable = true)
 |-- product_id: integer (nullable = true)
 |-- unit_price: double (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- discount: double (nullable = true)



In [23]:
df_employees.printSchema()

root
 |-- employee_id: integer (nullable = true)
 |-- last_name: string (nullable = true)
 |-- first_name: string (nullable = true)
 |-- title: string (nullable = true)
 |-- title_of_courtesy: string (nullable = true)
 |-- birth_date: date (nullable = true)
 |-- hire_date: date (nullable = true)
 |-- address: string (nullable = true)
 |-- city: string (nullable = true)
 |-- region: string (nullable = true)
 |-- postal_code: string (nullable = true)
 |-- country: string (nullable = true)
 |-- home_phone: string (nullable = true)
 |-- extension: integer (nullable = true)
 |-- photo: string (nullable = true)
 |-- notes: string (nullable = true)
 |-- reports_to: integer (nullable = true)
 |-- photo_path: string (nullable = true)



In [24]:
df_customers.printSchema()

root
 |-- customer_id: string (nullable = true)
 |-- company_name: string (nullable = true)
 |-- contact_name: string (nullable = true)
 |-- contact_title: string (nullable = true)
 |-- address: string (nullable = true)
 |-- city: string (nullable = true)
 |-- region: string (nullable = true)
 |-- postal_code: string (nullable = true)
 |-- country: string (nullable = true)
 |-- phone: string (nullable = true)
 |-- fax: string (nullable = true)



In [25]:
df_categories.printSchema()

root
 |-- category_id: integer (nullable = true)
 |-- category_name: string (nullable = true)
 |-- description: string (nullable = true)
 |-- picture: string (nullable = true)



In [26]:
df_categories.count()

8

In [27]:
from pyspark.sql import Row

In [28]:
rows = [
    Row(id=1, name='categories', count=df_categories.count()),
    Row(id=2, name='customers', count=df_customers.count()),
    Row(id=3, name='employees', count=df_employees.count()),
    Row(id=4, name='order_details', count=df_order_details.count()),
    Row(id=5, name='orders', count=df_orders.count()),
    Row(id=6, name='products', count=df_products.count()),
    Row(id=7, name='shippers', count=df_shippers.count()),
    Row(id=8, name='suppliers', count=df_suppliers.count())
]
df = spark.createDataFrame(rows)
df.show()

+---+-------------+-----+
| id|         name|count|
+---+-------------+-----+
|  1|   categories|    8|
|  2|    customers|   91|
|  3|    employees|    9|
|  4|order_details| 2155|
|  5|       orders|  830|
|  6|     products|   77|
|  7|     shippers|    6|
|  8|    suppliers|   29|
+---+-------------+-----+



In [29]:
from pyspark.sql.functions import min
df_orders.select(min("freight")).show()

+------------+
|min(freight)|
+------------+
|        0.02|
+------------+



In [30]:
from pyspark.sql.functions import max
df_orders.select(max("freight")).show()

+------------+
|max(freight)|
+------------+
|     1007.64|
+------------+



In [31]:
from pyspark.sql.functions import mean
df_orders.select(mean("freight")).show()

+-----------------+
|     avg(freight)|
+-----------------+
|78.24420481927719|
+-----------------+



In [32]:
from pyspark.sql.functions import stddev
df_products.select(stddev("unit_price")).show()

+------------------+
|stddev(unit_price)|
+------------------+
| 33.82931122234885|
+------------------+



La lazy evaluation (évaluation paresseuse) est le principe suivant : Spark n’exécute pas immédiatement les opérations qu'on lui demande. Il construit d’abord un plan d’exécution, puis ne lance réellement les calculs que lorsqu’une action est appelée.

Par exemple, df_min_freight = df_orders.select(min("freight")) est une transformation, elle n'est pas exécutée immédiatement mais est intégrée dans un DAG (graphe de calcul). Ce n'est que lorsqu'une action du type df_min_freight.show() arrive que l'ensemble est exécuté selon le DAG. Ce processus permet d'optimiser tous les traitements inscrits dans le DAG avant leur exécution.

Les opérations du type select, filter, groupBy sont des transformations non exécutées immédiatement mais intégrées au DAG, les opérations show, count, write sont des actions qui déclenchent l'exécution du DAG.

Sous http://localhost:4040, on peut voir la liste des jobs et des stages. Pour chaque stage, on peut voir le DAG, les métriques associées et la liste des tasks correspondantes.
Un job est lancé sur une action, un job se décompose en plusieurs étapes (stages), la séparation entre stages correspondant principalement à un shuffle. Une étape est exécutée en parallèle sous forme de tasks. Donc un "stage" est un ensemble de "tasks" qui peuvent être exécutées parallèlement sans nécessiter de shuffle entre elles.

In [39]:
import pandas as pd

In [41]:
%time pd_orders = pd.read_csv("../data/order_details.csv")

CPU times: user 4.38 ms, sys: 0 ns, total: 4.38 ms
Wall time: 5.48 ms


In [42]:
%time df_order_details = spark.read.csv("../data/order_details.csv",header=True,inferSchema=True)

CPU times: user 0 ns, sys: 4.47 ms, total: 4.47 ms
Wall time: 210 ms


In [43]:
%time pd_sample = pd.read_csv("../data/csv_sample_file_10MB.csv")

CPU times: user 55.2 ms, sys: 5.76 ms, total: 61 ms
Wall time: 136 ms


In [44]:
%time df_sample = spark.read.csv("../data/csv_sample_file_10MB.csv")

CPU times: user 2.19 ms, sys: 2.48 ms, total: 4.67 ms
Wall time: 152 ms


Spark n'est pas plus pertinent que Pandas pour les petits fichiers, mais il est plus performant pour les gros fichiers.

In [45]:
df_orders.columns

['order_id',
 'customer_id',
 'employee_id',
 'order_date',
 'required_date',
 'shipped_date',
 'ship_via',
 'freight',
 'ship_name',
 'ship_address',
 'ship_city',
 'ship_region',
 'ship_postal_code',
 'ship_country']

In [46]:
df_orders.dtypes

[('order_id', 'int'),
 ('customer_id', 'string'),
 ('employee_id', 'int'),
 ('order_date', 'date'),
 ('required_date', 'date'),
 ('shipped_date', 'date'),
 ('ship_via', 'int'),
 ('freight', 'double'),
 ('ship_name', 'string'),
 ('ship_address', 'string'),
 ('ship_city', 'string'),
 ('ship_region', 'string'),
 ('ship_postal_code', 'string'),
 ('ship_country', 'string')]

In [47]:
df_orders.show(2)

+--------+-----------+-----------+----------+-------------+------------+--------+-------+--------------------+------------------+---------+-----------+----------------+------------+
|order_id|customer_id|employee_id|order_date|required_date|shipped_date|ship_via|freight|           ship_name|      ship_address|ship_city|ship_region|ship_postal_code|ship_country|
+--------+-----------+-----------+----------+-------------+------------+--------+-------+--------------------+------------------+---------+-----------+----------------+------------+
|   10248|      VINET|          5|1996-07-04|   1996-08-01|  1996-07-16|       3|  32.38|Vins et alcools C...|59 rue de l'Abbaye|    Reims|       NULL|           51100|      France|
|   10249|      TOMSP|          6|1996-07-05|   1996-08-16|  1996-07-10|       1|  11.61|  Toms Spezialit�ten|     Luisenstr. 48|  M�nster|       NULL|           44087|     Germany|
+--------+-----------+-----------+----------+-------------+------------+--------+-------+-

Il n'y a pas vraiment de types mals configurés, éventuellement freight qui pourrait être en DECIMAL(8,2), mais le gain en mémoire ne serait justifié que sur un gros volume dans une situation de système proche de ses capacités maximales. On éviterait aussi les erreurs d'arrondi (par ex 11.609998 au lieu de 11.61).